### ISSUE TITLE
too much magic

### ISSUE THREAD

**Issue by uniphil** [Calama/kale#1]
Models should define their schema. Explicit is better than implicit. Maybe kale should give some helpers there, and the collections stuff it's doing is good. But just grinding things through jsonpickle does _not_ produce nice schemas for non-trivial objects.

For example, lists are native json, but tuples are not. jsonpickle tries to help out by encoding some special metadata that tells it to restore the resulting json list as a tuple as you'd like, but this messes up the json structure in non-obvious ways.

``` python
>>> from datetime import datetime
>>> from kale import KaleModel
>>> class Thing(KaleModel):
...     def __init__(self, value):
...             self.versions = []
...             self.value = value
...     def value():
...             fget = lambda self: self.versions[0][1]
...             def fset(self, value):
...                     self.versions.insert(0, (datetime.now(), value))
...             return locals()
...     value = property(**value())
... 
>>> thing = Thing(4)
>>> thing.flat
{
    'py/object': '__main__.Thing',
    'versions': [
        {
            'py/tuple': [
                {
                    'py/repr': 'datetime/datetime.datetime(2013, 2, 28, 16, 49, 53, 878575)'
                },
                4
            ]
        }
    ]
}
>>> thing.value = 5
>>> thing.flat
{
    'py/object': '__main__.Thing',
    'versions': [
        {
            'py/tuple': [
                {
                    'py/repr': 'datetime/datetime.datetime(2013, 2, 28, 16, 50, 11, 364174)'
                },
                5
            ]
        },
        {
            'py/tuple': [
                {
                    'py/repr': 'datetime/datetime.datetime(2013, 2, 28, 16, 49, 53, 878575)'
                },
                4
            ]
        }
    ]
}
```

Horrible. To query by value, you'd have to do something like

``` python
Thing.find({'versions.py/tuple.1': 4})
```

If it had been a list, which is a native json structure, that query could be something more expected, like

``` python
Thing.find({'versions.1': 4})
```

actually... that wouldn't work because of an annoying bug for nested arrays in mongodb, but whatever. Point is, jsonpickle is generating unpredictable schemas, killing the utility of kale, and making things much more implicit than they should be.

Datetimes, which are actually handled natively by pymongo, have been stringified into something nearly unqueriable by jsonpickle. Just to access that string you'd need to use `{'versions.py/tuple.0.py/repr': '[something?...]'}`... ... and that gets you an awkward string.

Bottom line is that jsonpickle was seemingly not designed to generate a json document whose structure matches your object's structure, it just makes some json structure which can be restored back to your object.

Add to that the issues jsonpickle has with serializing pymongo `ObjectId`s, and I'm ready to ditch jsonpickle all together. Re-write it more simply and appropriately for the goals of kale. Eventually.

The pickle protocol still seems like an elegant way to handle object serialization. Kale should work fine as-is if `__getstate__` and `__setstate__` are defined right now, so that's going to be recommended practice and possibly enforced later.

**Comment by uniphil** [comment]
Ripping out jsonpickle. Not using the pickling protocol for anything at the moment. For serialization, make the Model an AttrDict (a la minimongo) -- directly saveable in mongo. Leave re-instantiation up to the user. Helpful, not magical.
